Issue #15769: added suppression for separate code blocks and updated documentation

Zopsss opened the pull request:
part of #15769

- added suppression
- updated documentation
- added test cases for separate code blocks

---

Diff Regression config: https://gist.githubusercontent.com/Zopsss/eef363a452104d034a9ce7be532f647d/raw/3a36ba1cad81f3dff628f24c1c9b1591a2c5ef78/master_google_checks.xml

Diff Regression patch config: https://gist.githubusercontent.com/Zopsss/4d1852aea0ab161ad746e43f9ee3362d/raw/eec7d9700e3b3b78fff0954566d392be6f937846/block_code_indentation_google_checks.xml

Diff Regression projects: https://gist.githubusercontent.com/romani/015aff9c00dd09b2b9db3f72951bc61f/raw/9891630875ee62e6f8a2975721841e8d80536d4d/Projects.properties

Too big:
https://gist.githubusercontent.com/Zopsss/22adadb570e4deb95296917244c580b3/raw/144d72ee70de11891eabe4a342cbb0da37153a20/projects-to-test-on-for-github-action.properties

romani commented on src/it/resources/com/google/checkstyle/test/chapter4formatting/rule412nonemptyblocks/InputRightCurlySwitchCasesBlocks.java:
```diff
@@ -95,13 +95,11 @@ public static void test4() {
+          {
                   int x = 0; // violation '.* incorrect indentation .*, expected .* 8, 10.'
-          } // violation '.* incorrect indentation level 10, expected level should be 8.'
+          }
```
I did not expect change in behavior in this file.
Please share AST for case/default, we might need to adjust xpath a bit to not suppress such cases.

Zopsss replied:
AST of:

```java
public class Test {
  public static void test4() {
    int mode = 0;
    switch (mode) {
      case 1: {
          int x = 1;
          break;
        } default:
          {
                  int x = 0;
          }
    }
  }
}
```

```
$ java -jar .\checkstyle-10.18.2-all.jar -t .\Test.java
COMPILATION_UNIT -> COMPILATION_UNIT [1:0]
`--CLASS_DEF -> CLASS_DEF [1:0]
    |--MODIFIERS -> MODIFIERS [1:0]
    |   `--LITERAL_PUBLIC -> public [1:0]
    |--LITERAL_CLASS -> class [1:7]
    |--IDENT -> Test [1:13]
    `--OBJBLOCK -> OBJBLOCK [1:18]
        |--LCURLY -> { [1:18]
        |--METHOD_DEF -> METHOD_DEF [2:2]
        |   |--MODIFIERS -> MODIFIERS [2:2]
        |   |   |--LITERAL_PUBLIC -> public [2:2]
        |   |   `--LITERAL_STATIC -> static [2:9]
        |   |--TYPE -> TYPE [2:16]
        |   |   `--LITERAL_VOID -> void [2:16]
        |   |--IDENT -> test4 [2:21]
        |   |--LPAREN -> ( [2:26]
        |   |--PARAMETERS -> PARAMETERS [2:27]
        |   |--RPAREN -> ) [2:27]
        |   `--SLIST -> { [2:29]
        |       |--VARIABLE_DEF -> VARIABLE_DEF [3:4]
        |       |   |--MODIFIERS -> MODIFIERS [3:4]
        |       |   |--TYPE -> TYPE [3:4]
        |       |   |   `--LITERAL_INT -> int [3:4]
        |       |   |--IDENT -> mode [3:8]
        |       |   `--ASSIGN -> = [3:13]
        |       |       `--EXPR -> EXPR [3:15]
        |       |           `--NUM_INT -> 0 [3:15]
        |       |--SEMI -> ; [3:16]
        |       |--LITERAL_SWITCH -> switch [4:4]
        |       |   |--LPAREN -> ( [4:11]
        |       |   |--EXPR -> EXPR [4:12]
        |       |   |   `--IDENT -> mode [4:12]
        |       |   |--RPAREN -> ) [4:16]
        |       |   |--LCURLY -> { [4:18]
        |       |   |--CASE_GROUP -> CASE_GROUP [5:6]
        |       |   |   |--LITERAL_CASE -> case [5:6]
        |       |   |   |   |--EXPR -> EXPR [5:11]
        |       |   |   |   |   `--NUM_INT -> 1 [5:11]
        |       |   |   |   `--COLON -> : [5:12]
        |       |   |   `--SLIST -> SLIST [5:14]
        |       |   |       `--SLIST -> { [5:14]
        |       |   |           |--VARIABLE_DEF -> VARIABLE_DEF [6:10]
        |       |   |           |   |--MODIFIERS -> MODIFIERS [6:10]
        |       |   |           |   |--TYPE -> TYPE [6:10]
        |       |   |           |   |   `--LITERAL_INT -> int [6:10]
        |       |   |           |   |--IDENT -> x [6:14]
        |       |   |           |   `--ASSIGN -> = [6:16]
        |       |   |           |       `--EXPR -> EXPR [6:18]
        |       |   |           |           `--NUM_INT -> 1 [6:18]
        |       |   |           |--SEMI -> ; [6:19]
        |       |   |           |--LITERAL_BREAK -> break [7:10]
        |       |   |           |   `--SEMI -> ; [7:15]
        |       |   |           `--RCURLY -> } [8:8]
        |       |   |--CASE_GROUP -> CASE_GROUP [8:10]
        |       |   |   |--LITERAL_DEFAULT -> default [8:10]
        |       |   |   |   `--COLON -> : [8:17]
        |       |   |   `--SLIST -> SLIST [9:10]
        |       |   |       `--SLIST -> { [9:10]
        |       |   |           |--VARIABLE_DEF -> VARIABLE_DEF [10:18]
        |       |   |           |   |--MODIFIERS -> MODIFIERS [10:18]
        |       |   |           |   |--TYPE -> TYPE [10:18]
        |       |   |           |   |   `--LITERAL_INT -> int [10:18]
        |       |   |           |   |--IDENT -> x [10:22]
        |       |   |           |   `--ASSIGN -> = [10:24]
        |       |   |           |       `--EXPR -> EXPR [10:26]
        |       |   |           |           `--NUM_INT -> 0 [10:26]
        |       |   |           |--SEMI -> ; [10:27]
        |       |   |           `--RCURLY -> } [11:10]
        |       |   `--RCURLY -> } [12:4]
        |       `--RCURLY -> } [13:2]
        `--RCURLY -> } [14:0]
```

The violation was a false-positive so I think this is good that we were able to suppress it. I guess there's no need to change xpath

romani commented:
```
$ cat Test.java 
public class Test {
    {}
  public static void test4() {
    {}
    int mode = 0;
    switch (mode) {
      case 1: {
          int x = 1;
          break;
        } default:
          {
                  int x = 0;
          }
    }
  }
}

$ java -jar checkstyle-10.18.3-SNAPSHOT-all.jar -t Test.java 
COMPILATION_UNIT -> COMPILATION_UNIT [1:0]
`--CLASS_DEF -> CLASS_DEF [1:0]
    |--MODIFIERS -> MODIFIERS [1:0]
    |   `--LITERAL_PUBLIC -> public [1:0]
    |--LITERAL_CLASS -> class [1:7]
    |--IDENT -> Test [1:13]
    `--OBJBLOCK -> OBJBLOCK [1:18]
        |--LCURLY -> { [1:18]
        |--INSTANCE_INIT -> INSTANCE_INIT [2:4]
        |   `--SLIST -> { [2:4]
        |       `--RCURLY -> } [2:5]
        |--METHOD_DEF -> METHOD_DEF [3:2]
        |   |--MODIFIERS -> MODIFIERS [3:2]
        |   |   |--LITERAL_PUBLIC -> public [3:2]
        |   |   `--LITERAL_STATIC -> static [3:9]
        |   |--TYPE -> TYPE [3:16]
        |   |   `--LITERAL_VOID -> void [3:16]
        |   |--IDENT -> test4 [3:21]
        |   |--LPAREN -> ( [3:26]
        |   |--PARAMETERS -> PARAMETERS [3:27]
        |   |--RPAREN -> ) [3:27]
        |   `--SLIST -> { [3:29]
        |       |--SLIST -> { [4:4]
        |       |   `--RCURLY -> } [4:5]
        |       |--VARIABLE_DEF -> VARIABLE_DEF [5:4]
        |       |   |--MODIFIERS -> MODIFIERS [5:4]
        |       |   |--TYPE -> TYPE [5:4]
        |       |   |   `--LITERAL_INT -> int [5:4]
        |       |   |--IDENT -> mode [5:8]
        |       |   `--ASSIGN -> = [5:13]
        |       |       `--EXPR -> EXPR [5:15]
        |       |           `--NUM_INT -> 0 [5:15]
        |       |--SEMI -> ; [5:16]
        |       |--LITERAL_SWITCH -> switch [6:4]
        |       |   |--LPAREN -> ( [6:11]
        |       |   |--EXPR -> EXPR [6:12]
        |       |   |   `--IDENT -> mode [6:12]
        |       |   |--RPAREN -> ) [6:16]
        |       |   |--LCURLY -> { [6:18]
        |       |   |--CASE_GROUP -> CASE_GROUP [7:6]
        |       |   |   |--LITERAL_CASE -> case [7:6]
        |       |   |   |   |--EXPR -> EXPR [7:11]
        |       |   |   |   |   `--NUM_INT -> 1 [7:11]
        |       |   |   |   `--COLON -> : [7:12]
        |       |   |   `--SLIST -> SLIST [7:14]
        |       |   |       `--SLIST -> { [7:14]
        |       |   |           |--VARIABLE_DEF -> VARIABLE_DEF [8:10]
        |       |   |           |   |--MODIFIERS -> MODIFIERS [8:10]
        |       |   |           |   |--TYPE -> TYPE [8:10]
        |       |   |           |   |   `--LITERAL_INT -> int [8:10]
        |       |   |           |   |--IDENT -> x [8:14]
        |       |   |           |   `--ASSIGN -> = [8:16]
        |       |   |           |       `--EXPR -> EXPR [8:18]
        |       |   |           |           `--NUM_INT -> 1 [8:18]
        |       |   |           |--SEMI -> ; [8:19]
        |       |   |           |--LITERAL_BREAK -> break [9:10]
        |       |   |           |   `--SEMI -> ; [9:15]
        |       |   |           `--RCURLY -> } [10:8]
        |       |   |--CASE_GROUP -> CASE_GROUP [10:10]
        |       |   |   |--LITERAL_DEFAULT -> default [10:10]
        |       |   |   |   `--COLON -> : [10:17]
        |       |   |   `--SLIST -> SLIST [11:10]
        |       |   |       `--SLIST -> { [11:10]
        |       |   |           |--VARIABLE_DEF -> VARIABLE_DEF [12:18]
        |       |   |           |   |--MODIFIERS -> MODIFIERS [12:18]
        |       |   |           |   |--TYPE -> TYPE [12:18]
        |       |   |           |   |   `--LITERAL_INT -> int [12:18]
        |       |   |           |   |--IDENT -> x [12:22]
        |       |   |           |   `--ASSIGN -> = [12:24]
        |       |   |           |       `--EXPR -> EXPR [12:26]
        |       |   |           |           `--NUM_INT -> 0 [12:26]
        |       |   |           |--SEMI -> ; [12:27]
        |       |   |           `--RCURLY -> } [13:10]
        |       |   `--RCURLY -> } [14:4]
        |       `--RCURLY -> } [15:2]
        `--RCURLY -> } [16:0]

$ java -jar checkstyle-10.18.3-SNAPSHOT-all.jar -b "//SLIST[not(parent::CASE_GROUP)]/SLIST | //SLIST[not(parent::CASE_GROUP)]/SLIST/RCURLY" Test.java 
COMPILATION_UNIT -> COMPILATION_UNIT [1:0]
`--CLASS_DEF -> CLASS_DEF [1:0]
    `--OBJBLOCK -> OBJBLOCK [1:18]
        |--METHOD_DEF -> METHOD_DEF [3:2]
        |   `--SLIST -> { [3:29]
        |       |--SLIST -> { [4:4]
---------
COMPILATION_UNIT -> COMPILATION_UNIT [1:0]
`--CLASS_DEF -> CLASS_DEF [1:0]
    `--OBJBLOCK -> OBJBLOCK [1:18]
        |--METHOD_DEF -> METHOD_DEF [3:2]
        |   `--SLIST -> { [3:29]
        |       |--SLIST -> { [4:4]
        |       |   `--RCURLY -> } [4:5]
```

romani commented:
@Zopsss , please put `//SLIST[not(parent::CASE_GROUP)]/SLIST | //SLIST[not(parent::CASE_GROUP)]/SLIST/RCURLY` in xpath and lets see how it work on large scale.

Zopsss commented:
> @Zopsss , please put `//SLIST[not(parent::CASE_GROUP)]/SLIST | //SLIST[not(parent::CASE_GROUP)]/SLIST/RCURLY` in xpath and lets see how it work on large scale.

Done

romani commented on src/main/resources/google_checks.xml:
```diff
@@ -285,6 +285,13 @@
       <property name="arrayInitIndent" value="2"/>
     </module>
+    <!-- Suppression for block code until we find a way to detect them properly
+         https://github.com/checkstyle/checkstyle/issues/15769 -->
```
Please put until word before link, link and such word should be on same line.
This is how we detect closed issues.

Zopsss replied:
done

romani reviewed (changes requested):
Last:

romani commented:
Please generate diff report

Zopsss commented:
Github, generate report

romani commented:
Github, generate report for configs in PR description

romani reviewed (approved):
ok to merge if dff report is good.

romani commented:
`Caused by: com.puppycrawl.tools.checkstyle.api.CheckstyleException: IllegalStateException occurred while parsing file /home/runner/work/checkstyle/checkstyle/.ci-temp/src/main/java/checkstyle/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/javadocpropertiesgenerator/InputJavadocPropertiesGeneratorParseError.java.`

we need to recehck Checkstyle excludes to avoid this file.

romani commented:
Github, generate report for configs in PR description

romani commented:
>java.lang.OutOfMemoryError: Java heap space

Please reduce amount of projects, and return.

Zopsss commented:
Github, generate report for configs in PR description

romani commented:
>java.lang.OutOfMemoryError

Cut even more

romani commented:
@Zopsss , update link `Diff Regression projects:` to put link to gist with less projects in it.

Zopsss commented:
Github, generate report

Zopsss commented:
GitHub, generate report

Zopsss commented:
removed openjdk17 from the list, it was causing the heap error and also added 

```xml
    <!-- as we run on regression even on non-compiled files we need to skip exceptions on them -->
    <property name="skipFileOnJavaParseException" value="true"/>
    <property name="javaParseExceptionSeverity" value="ignore"/>
```

to configs to skip exceptions.

Now it should generate the report successfully.

Zopsss commented:
Github, generate site

romani commented:
Github, generate report for configs in PR description

Zopsss commented:
report looks good to me and Roman. @rdiachenko Please do final review.

romani reviewed (approved):
Most valuable is guava diff, all looks good.

All other are less relevant, not I see no problems.

Zopsss commented:
ping @rdiachenko

rdiachenko reviewed (approved):
lgtm
